# CSE476 CA1 Project 1: Personal Budget Assistant Agent
**Domain:** Finance | **Topic:** T1. Personal Budget Assistant

--- 
### Agent Architecture & Features:
1. **Plan-Act ReAct Loop:** Multi-step reasoning where the agent formulates thoughts, decides tool calls, receives observations, and computes decisions.
2. **Tools Called:**
   - `add_expense(item, amount, category)`: Logs an expense and updates category balance.
   - `get_summary(category)`: Retrieves total budget, spent funds, and remaining balance.
   - `set_savings_goal(target_amount, month)` *(Group of 3 Add-on)*: Sets a savings goal target.
   - `check_affordability(expense_name, cost)`: Calculates whether a purchase/trip fits remaining budget and savings targets.
3. **Memory:** Retains state (expenses logged, savings targets, category totals) and multi-turn conversation history.

In [1]:
# Initialize environment and import agent components
from memory import BudgetMemory
from tools import reset_memory
from agent import PersonalBudgetAgent

# Reset global memory state with a student monthly budget of $5,000
memory = reset_memory(initial_budget=5000.0)
agent = PersonalBudgetAgent(memory=memory, verbose=True)
print("✅ Personal Budget Assistant Agent initialized.")

✅ Personal Budget Assistant Agent initialized.


## Turn 1: Logging Multiple Expenses (Multi-step Tool Execution)
**User Goal:** *"Log $150 spent on textbooks and $80 on groceries"*

In [2]:
goal_1 = "Log $150 spent on textbooks and $80 on groceries"
response_1 = agent.run(goal_1)


🤖 AGENT GOAL: 'Log $150 spent on textbooks and $80 on groceries'

[Step 1]
  💭 Thought: I need to log the expense 'Textbooks' of $80.0 in category 'shopping'.
  🛠️ Action: add_expense
  📥 Tool Input: {"item": "Textbooks", "amount": 80.0, "category": "shopping"}
  🔍 Observation: {
  "status": "success",
  "message": "Logged expense 'Textbooks' of $80.00 under category 'shopping'.",
  "logged_entry": {
    "item": "Textbooks",
    "amount": 80.0,
    "category": "shopping",
    "timestamp": "2026-08-31 14:21"
  },
  "category_total": 80.0,
  "category_limit": 500.0,
  "is_overspent": false,
  "remaining_balance": 4920.0
}

[Step 2]
  💭 Thought: I need to log the expense 'Groceries' of $150.0 in category 'food'.
  🛠️ Action: add_expense
  📥 Tool Input: {"item": "Groceries", "amount": 150.0, "category": "food"}
  🔍 Observation: {
  "status": "success",
  "message": "Logged expense 'Groceries' of $150.00 under category 'food'.",
  "logged_entry": {
    "item": "Groceries",
    "amount": 15

## Turn 2: Setting a Monthly Savings Goal (Group of 3 Add-on)
**User Goal:** *"Set a savings goal of $1000 for August"*

In [3]:
goal_2 = "Set a savings goal of 1000 for August"
response_2 = agent.run(goal_2)


🤖 AGENT GOAL: 'Set a savings goal of 1000 for August'

[Step 1]
  💭 Thought: Setting a monthly savings goal of $1000.0 for August.
  🛠️ Action: set_savings_goal
  📥 Tool Input: {"target_amount": 1000.0, "month": "August"}
  🔍 Observation: {
  "status": "success",
  "message": "Savings goal of $1000.00 set for August.",
  "target_amount": 1000.0,
  "month": "August",
  "current_remaining_balance": 4770.0,
  "balance_after_savings": 3770.0,
  "is_achievable": true
}

[Step 2]
  💭 Thought: Checking current budget summary to see if the savings goal is achievable with remaining funds.
  🛠️ Action: get_summary
  📥 Tool Input: {}
  🔍 Observation: {
  "scope": "overall",
  "monthly_budget": 5000.0,
  "total_spent": 230.0,
  "remaining_balance": 4770.0,
  "savings_goal": {
    "target": 1000.0,
    "month": "August"
  },
  "category_breakdown": {
    "shopping": 80.0,
    "food": 150.0
  },
  "overspent_categories": [],
  "total_expenses_logged": 2
}

  💡 Final Answer / Decision:
  Savings goa

## Turn 3: Agentic Decision & Affordability Check
**User Goal:** *"Can I afford a $2000 trip?"*

*Notice how the agent checks current remaining balance ($4,770), factors in the savings goal ($1,000), calls the `check_affordability` tool, and decides mathematically instead of just generating text.*

In [4]:
goal_3 = "Can I afford a 2000 trip?"
response_3 = agent.run(goal_3)


🤖 AGENT GOAL: 'Can I afford a 2000 trip?'

[Step 1]
  💭 Thought: First, I must inspect our current remaining budget and savings targets via get_summary.
  🛠️ Action: get_summary
  📥 Tool Input: {}
  🔍 Observation: {
  "scope": "overall",
  "monthly_budget": 5000.0,
  "total_spent": 230.0,
  "remaining_balance": 4770.0,
  "savings_goal": {
    "target": 1000.0,
    "month": "August"
  },
  "category_breakdown": {
    "shopping": 80.0,
    "food": 150.0
  },
  "overspent_categories": [],
  "total_expenses_logged": 2
}

[Step 2]
  💭 Thought: Now I will evaluate whether spending $2000.0 on 'Weekend Trip' fits within our remaining budget.
  🛠️ Action: check_affordability
  📥 Tool Input: {"expense_name": "Weekend Trip", "cost": 2000.0}
  🔍 Observation: {
  "expense_name": "Weekend Trip",
  "cost": 2000.0,
  "current_balance": 4770.0,
  "balance_after_purchase": 2770.0,
  "savings_target": 1000.0,
  "verdict": "AFFORDABLE",
  "explanation": "Yes! The purchase 'Weekend Trip' ($2000.00) is aff

## Memory Proof: Turn History & State Read-Back
Demonstrating that the agent retains memory across all turns in the session.

In [5]:
print("=== CONVERSATION HISTORY MEMORY ===")
print(agent.memory.get_conversation_context())

print("
=== STRUCTURED BUDGET STATE MEMORY ===")
import json
print(json.dumps(agent.memory.get_summary(), indent=2))

SyntaxError: unterminated string literal (detected at line 4) (224059207.py, line 4)